File: modelingassistant/pythonclient/mistaketypes.py
"""
This file contains all mistake types and categories.
"""

from constants import LEARNING_CORPUS_PATH
from corpus import corpus as runtime_corpus
from fileserdes import load_lc
from learningcorpus import MistakeTypeCategory, MistakeType
from utils import mdf

corpus = load_lc(LEARNING_CORPUS_PATH)

# Populate dictionaries
MISTAKE_TYPE_CATEGORIES_BY_NAME: dict[str, MistakeTypeCategory] = {c.name: c for c in corpus.mistakeTypeCategories}
MISTAKE_TYPES_BY_NAME: dict[str, MistakeType] = {mt.name: mt for mt in corpus.mistakeTypes()}

# Short-name references to the above dicts for greater code legibility
_MTCS = MISTAKE_TYPE_CATEGORIES_BY_NAME
_MTS = MISTAKE_TYPES_BY_NAME

# Add mistake detection format attribute to mistake types in _MTS
for mt in runtime_corpus.mistakeTypes():
    _MTS[mt.name].md_format = getattr(mt, "md_format", mdf([], []))


# Mistake type categories
CLASS_MISTAKES: MistakeTypeCategory = _MTCS["Class mistakes"]
ATTRIBUTE_MISTAKES: MistakeTypeCategory = _MTCS["Attribute mistakes"]
RELATIONSHIP_MISTAKES: MistakeTypeCategory = _MTCS["Relationship mistakes"]
DESIGN_PATTERN_MISTAKES: MistakeTypeCategory = _MTCS["Design pattern mistakes"]
CLASS_NAME_MISTAKES: MistakeTypeCategory = _MTCS["Class name mistakes"]
ENUMERATION_MISTAKES: MistakeTypeCategory = _MTCS["Enumeration mistakes"]
ATTRIBUTE_NAME_MISTAKES: MistakeTypeCategory = _MTCS["Attribute name mistakes"]
ATTRIBUTE_IN_WRONG_CLASS_MISTAKES: MistakeTypeCategory = _MTCS["Attribute in wrong class mistakes"]
EXTRA_ATTRIBUTE_MISTAKES: MistakeTypeCategory = _MTCS["Extra attribute mistakes"]
MISSING_ASSOCIATION_AGGREGATION_MISTAKES: MistakeTypeCategory = _MTCS["Missing association/aggregation mistakes"]
EXTRA_ASSOCIATION_MISTAKES: MistakeTypeCategory = _MTCS["Extra association mistakes"]
MULTIPLICITY_MISTAKES: MistakeTypeCategory = _MTCS["Multiplicity mistakes"]
ROLE_NAME_MISTAKES: MistakeTypeCategory = _MTCS["Role name mistakes"]
ASSOCIATION_TYPE_MISTAKES: MistakeTypeCategory = _MTCS["Association type mistakes"]
ASSOCIATION_NAME_MISTAKES: MistakeTypeCategory = _MTCS["Association name mistakes"]
ASSOCIATION_CLASS_MISTAKES: MistakeTypeCategory = _MTCS["Association class mistakes"]
COMPOSITION_MISTAKES: MistakeTypeCategory = _MTCS["Composition mistakes"]
GENERALIZATION_MISTAKES: MistakeTypeCategory = _MTCS["Generalization mistakes"]
PLAYER_ROLE_PATTERN_MISTAKES: MistakeTypeCategory = _MTCS["Player-Role Pattern mistakes"]
ABSTRACTION_OCCURRENCE_PATTERN_MISTAKES: MistakeTypeCategory = _MTCS["Abstraction-Occurrence pattern mistakes"]

# Mistake types
MISSING_CLASS: MistakeType = _MTS["Missing class"]
EXTRA_CLASS: MistakeType = _MTS["Extra class"]
MISSING_ATTRIBUTE: MistakeType = _MTS["Missing attribute"]
WRONG_ATTRIBUTE_TYPE: MistakeType = _MTS["Wrong attribute type"]
MISSING_ATTRIBUTE_TYPE: MistakeType = _MTS["Missing attribute type"]
ATTRIBUTE_SHOULD_BE_STATIC: MistakeType = _MTS["Attribute should be static"]
ATTRIBUTE_SHOULD_NOT_BE_STATIC: MistakeType = _MTS["Attribute should not be static"]
PLURAL_CLASS_NAME: MistakeType = _MTS["Plural class name"]
LOWERCASE_CLASS_NAME: MistakeType = _MTS["Lowercase class name"]
SOFTWARE_ENGINEERING_TERM: MistakeType = _MTS["Software engineering term"]
BAD_CLASS_NAME_SPELLING: MistakeType = _MTS["Bad class name spelling"]
WRONG_CLASS_NAME: MistakeType = _MTS["Wrong class name"]
CLASS_SHOULD_BE_ENUM: MistakeType = _MTS["Class should be enum"]
ENUM_SHOULD_BE_CLASS: MistakeType = _MTS["Enum should be class"]
MISSING_ENUM: MistakeType = _MTS["Missing enum"]
EXTRA_ENUM: MistakeType = _MTS["Extra enum"]
BAD_ENUM_NAME_SPELLING: MistakeType = _MTS["Bad enum name spelling"]
MISSING_ENUM_ITEM: MistakeType = _MTS["Missing enum item"]
EXTRA_ENUM_ITEM: MistakeType = _MTS["Extra enum item"]
BAD_ENUM_ITEM_SPELLING: MistakeType = _MTS["Bad enum item spelling"]
BAD_ATTRIBUTE_NAME_SPELLING: MistakeType = _MTS["Bad attribute name spelling"]
UPPERCASE_ATTRIBUTE_NAME: MistakeType = _MTS["Uppercase attribute name"]
ATTRIBUTE_MISPLACED: MistakeType = _MTS["Attribute misplaced"]
ATTRIBUTE_DUPLICATED: MistakeType = _MTS["Attribute duplicated"]
ATTRIBUTE_MISPLACED_IN_GENERALIZATION_HIERARCHY: MistakeType = _MTS["Attribute misplaced in generalization hierarchy"]
PLURAL_ATTRIBUTE: MistakeType = _MTS["Plural attribute"]
LIST_ATTRIBUTE: MistakeType = _MTS["List attribute"]
EXTRA_ATTRIBUTE: MistakeType = _MTS["Extra attribute"]
MISSING_ASSOCIATION: MistakeType = _MTS["Missing association"]
MISSING_AGGREGATION: MistakeType = _MTS["Missing aggregation"]
MISSING_N_ARY_ASSOCIATION: MistakeType = _MTS["Missing n-ary association"]
USING_ATTRIBUTE_INSTEAD_OF_ASSOC: MistakeType = _MTS["Using attribute instead of assoc"]
EXTRA_ASSOCIATION: MistakeType = _MTS["Extra association"]
EXTRA_AGGREGATION: MistakeType = _MTS["Extra aggregation"]
EXTRA_N_ARY_ASSOCIATION: MistakeType = _MTS["Extra n-ary association"]
INFINITE_RECURSIVE_DEPENDENCY: MistakeType = _MTS["Infinite recursive dependency"]
WRONG_MULTIPLICITY: MistakeType = _MTS["Wrong multiplicity"]
MISSING_MULTIPLICITY: MistakeType = _MTS["Missing multiplicity"]
MISSING_ROLE_NAMES: MistakeType = _MTS["Missing role names"]
ROLE_SHOULD_BE_STATIC: MistakeType = _MTS["Role should be static"]
ROLE_SHOULD_NOT_BE_STATIC: MistakeType = _MTS["Role should not be static"]
BAD_ROLE_NAME_SPELLING: MistakeType = _MTS["Bad role name spelling"]
REPRESENTING_ACTION_WITH_ASSOC: MistakeType = _MTS["Representing action with assoc"]
WRONG_ROLE_NAME: MistakeType = _MTS["Wrong role name"]
USING_AGGREGATION_INSTEAD_OF_ASSOC: MistakeType = _MTS["Using aggregation instead of assoc"]
USING_COMPOSITION_INSTEAD_OF_ASSOC: MistakeType = _MTS["Using composition instead of assoc"]
USING_DIRECTED_RELATIONSHIP_INSTEAD_OF_UNDIRECTED: MistakeType = \
    _MTS["Using directed relationship instead of undirected"]
USING_UNDIRECTED_RELATIONSHIP_INSTEAD_OF_DIRECTED: MistakeType = \
    _MTS["Using undirected relationship instead of directed"]
WRONG_RELATIONSHIP_DIRECTION: MistakeType = _MTS["Wrong relationship direction"]
USING_COMPOSITION_INSTEAD_OF_AGGREGATION: MistakeType = _MTS["Using composition instead of aggregation"]
USING_BINARY_ASSOC_INSTEAD_OF_N_ARY_ASSOC: MistakeType = _MTS["Using binary assoc instead of n-ary assoc"]
USING_N_ARY_ASSOC_INSTEAD_OF_BINARY_ASSOC: MistakeType = _MTS["Using n-ary assoc instead of binary assoc"]
USING_INTERMEDIATE_CLASS_INSTEAD_OF_N_ARY_ASSOC: MistakeType = _MTS["Using intermediate class instead of n-ary assoc"]
USING_N_ARY_ASSOC_INSTEAD_OF_INTERMEDIATE_CLASS: MistakeType = _MTS["Using n-ary assoc instead of intermediate class"]
MISSING_ASSOCIATION_NAME: MistakeType = _MTS["Missing association name"]
BAD_ASSOCIATION_NAME_SPELLING: MistakeType = _MTS["Bad association name spelling"]
MISSING_ASSOC_CLASS: MistakeType = _MTS["Missing assoc class"]
EXTRA_ASSOC_CLASS: MistakeType = _MTS["Extra assoc class"]
BAD_ASSOC_CLASS_NAME_SPELLING: MistakeType = _MTS["Bad assoc class name spelling"]
ASSOC_CLASS_SHOULD_BE_CLASS: MistakeType = _MTS["Assoc class should be class"]
CLASS_SHOULD_BE_ASSOC_CLASS: MistakeType = _MTS["Class should be assoc class"]
MISSING_COMPOSITION: MistakeType = _MTS["Missing composition"]
EXTRA_COMPOSITION: MistakeType = _MTS["Extra composition"]
USING_ASSOC_INSTEAD_OF_AGGREGATION: MistakeType = _MTS["Using assoc instead of aggregation"]
USING_ASSOC_INSTEAD_OF_COMPOSITION: MistakeType = _MTS["Using assoc instead of composition"]
USING_AGGREGATION_INSTEAD_OF_COMPOSITION: MistakeType = _MTS["Using aggregation instead of composition"]
COMPOSED_PART_CONTAINED_IN_MORE_THAN_ONE_PARENT: MistakeType = _MTS["Composed part contained in more than one parent"]
INCOMPLETE_CONTAINMENT_TREE: MistakeType = _MTS["Incomplete containment tree"]
MISSING_GENERALIZATION: MistakeType = _MTS["Missing generalization"]
EXTRA_GENERALIZATION: MistakeType = _MTS["Extra generalization"]
GENERALIZATION_INAPPLICABLE: MistakeType = _MTS["Generalization inapplicable"]
SUBCLASS_NOT_DISTINCT_ACROSS_LIFETIME: MistakeType = _MTS["Subclass not distinct across lifetime"]
INHERITED_FEATURE_DOES_NOT_MAKE_SENSE_FOR_SUBCLASS: MistakeType = \
    _MTS["Inherited feature does not make sense for subclass"]
SUBCLASS_IS_AN_INSTANCE_OF_SUPERCLASS: MistakeType = _MTS["Subclass is an instance of superclass"]
NON_DIFFERENTIATED_SUBCLASS: MistakeType = _MTS["Non-differentiated subclass"]
WRONG_GENERALIZATION_DIRECTION: MistakeType = _MTS["Wrong generalization direction"]
WRONG_SUPERCLASS: MistakeType = _MTS["Wrong superclass"]
MISSING_PR_PATTERN: MistakeType = _MTS["Missing PR pattern"]
INCOMPLETE_PR_PATTERN: MistakeType = _MTS["Incomplete PR pattern"]
SUBCLASS_SHOULD_BE_FULL_PR_PATTERN: MistakeType = _MTS["Subclass should be full PR pattern"]
SUBCLASS_SHOULD_BE_ASSOC_PR_PATTERN: MistakeType = _MTS["Subclass should be assoc PR pattern"]
SUBCLASS_SHOULD_BE_ENUM_PR_PATTERN: MistakeType = _MTS["Subclass should be enum PR pattern"]
ASSOC_SHOULD_BE_FULL_PR_PATTERN: MistakeType = _MTS["Assoc should be full PR pattern"]
ASSOC_SHOULD_BE_SUBCLASS_PR_PATTERN: MistakeType = _MTS["Assoc should be subclass PR pattern"]
ASSOC_SHOULD_BE_ENUM_PR_PATTERN: MistakeType = _MTS["Assoc should be enum PR pattern"]
ENUM_SHOULD_BE_FULL_PR_PATTERN: MistakeType = _MTS["Enum should be full PR pattern"]
ENUM_SHOULD_BE_SUBCLASS_PR_PATTERN: MistakeType = _MTS["Enum should be subclass PR pattern"]
ENUM_SHOULD_BE_ASSOC_PR_PATTERN: MistakeType = _MTS["Enum should be assoc PR pattern"]
FULL_PR_PATTERN_SHOULD_BE_SUBCLASS: MistakeType = _MTS["Full PR pattern should be subclass"]
FULL_PR_PATTERN_SHOULD_BE_ASSOC: MistakeType = _MTS["Full PR pattern should be assoc"]
FULL_PR_PATTERN_SHOULD_BE_ENUM: MistakeType = _MTS["Full PR pattern should be enum"]
MISSING_AO_PATTERN: MistakeType = _MTS["Missing AO pattern"]
INCOMPLETE_AO_PATTERN: MistakeType = _MTS["Incomplete AO pattern"]
GENERALIZATION_SHOULD_BE_ASSOC_AO_PATTERN: MistakeType = _MTS["Generalization should be assoc AO pattern"]

File: modelingassistant/src/learningcorpus/mistaketypes/MistakeTypes.java
package learningcorpus.mistaketypes;

import java.util.HashMap;
import java.util.Map;
import learningcorpus.LearningCorpus;
import learningcorpus.MistakeType;
import learningcorpus.MistakeTypeCategory;

/**
 * This class contains all mistake types and categories.
 */
public class MistakeTypes {

  /** The path of the learning corpus instance with mistake types. */
  public static final String LEARNING_CORPUS_PATH =
      "../modelingassistant/learningcorpusinstances/default.learningcorpus";

  /** Map of mistake type categories by name. */
  public static final Map<String, MistakeTypeCategory> MISTAKE_TYPE_CATEGORIES_BY_NAME = new HashMap<>();

  /** Map of mistake types by name. */
  public static final Map<String, MistakeType> MISTAKE_TYPES_BY_NAME = new HashMap<>();

  // Short-name references to the above maps for greater code legibility
  private static final Map<String, MistakeTypeCategory> MTCS = MISTAKE_TYPE_CATEGORIES_BY_NAME;
  private static final Map<String, MistakeType> MTS = MISTAKE_TYPES_BY_NAME;

  static {
    var learningCorpus = LearningCorpus.fromFile(LEARNING_CORPUS_PATH);
    learningCorpus.getMistakeTypeCategories().forEach(mtc -> MISTAKE_TYPE_CATEGORIES_BY_NAME.put(mtc.getName(), mtc));
    learningCorpus.getMistakeTypes().forEach(mt -> MISTAKE_TYPES_BY_NAME.put(mt.getName(), mt));
  }

  // Mistake type categories

  /** The category for class mistakes. */
  public static final MistakeTypeCategory CLASS_MISTAKES = MTCS.get("Class mistakes");

  /** The category for attribute mistakes. */
  public static final MistakeTypeCategory ATTRIBUTE_MISTAKES = MTCS.get("Attribute mistakes");

  /** The category for relationship mistakes. */
  public static final MistakeTypeCategory RELATIONSHIP_MISTAKES = MTCS.get("Relationship mistakes");

  /** The category for design pattern mistakes. */
  public static final MistakeTypeCategory DESIGN_PATTERN_MISTAKES = MTCS.get("Design pattern mistakes");

  /** The category for class name mistakes. */
  public static final MistakeTypeCategory CLASS_NAME_MISTAKES = MTCS.get("Class name mistakes");

  /** The category for enumeration mistakes. */
  public static final MistakeTypeCategory ENUMERATION_MISTAKES = MTCS.get("Enumeration mistakes");

  /** The category for extra attribute mistakes. */
  public static final MistakeTypeCategory EXTRA_ATTRIBUTE_MISTAKES = MTCS.get("Extra attribute mistakes");

  /** The category for attribute name mistakes. */
  public static final MistakeTypeCategory ATTRIBUTE_NAME_MISTAKES = MTCS.get("Attribute name mistakes");

  /** The category for attribute in wrong class mistakes. */
  public static final MistakeTypeCategory ATTRIBUTE_IN_WRONG_CLASS_MISTAKES =
      MTCS.get("Attribute in wrong class mistakes");

  /** The category for missing association mistakes. */
  public static final MistakeTypeCategory MISSING_ASSOCIATION_MISTAKES = MTCS.get("Missing association mistakes");

  /** The category for extra association mistakes. */
  public static final MistakeTypeCategory EXTRA_ASSOCIATION_MISTAKES = MTCS.get("Extra association mistakes");

  /** The category for association type mistakes. */
  public static final MistakeTypeCategory ASSOCIATION_TYPE_MISTAKES = MTCS.get("Association type mistakes");

  /** The category for association name mistakes. */
  public static final MistakeTypeCategory ASSOCIATION_NAME_MISTAKES = MTCS.get("Association name mistakes");

  /** The category for multiplicity mistakes. */
  public static final MistakeTypeCategory MULTIPLICITY_MISTAKES = MTCS.get("Multiplicity mistakes");

  /** The category for role name mistakes. */
  public static final MistakeTypeCategory ROLE_NAME_MISTAKES = MTCS.get("Role name mistakes");

  /** The category for association class mistakes. */
  public static final MistakeTypeCategory ASSOCIATION_CLASS_MISTAKES = MTCS.get("Association class mistakes");

  /** The category for association mistakes. */
  public static final MistakeTypeCategory ASSOCIATION_MISTAKES = MTCS.get("Association mistakes");

  /** The category for composition mistakes. */
  public static final MistakeTypeCategory COMPOSITION_MISTAKES = MTCS.get("Composition mistakes");

  /** The category for generalization mistakes. */
  public static final MistakeTypeCategory GENERALIZATION_MISTAKES = MTCS.get("Generalization mistakes");

  /** The category for using different player-role pattern. */
  public static final MistakeTypeCategory USING_DIFFERENT_PLAYER_ROLE_PATTERN =
      MTCS.get("Using different Player-Role pattern");

  /** The category for player-role pattern mistakes. */
  public static final MistakeTypeCategory PLAYER_ROLE_PATTERN_MISTAKES = MTCS.get("Player-Role Pattern mistakes");

  /** The category for abstraction-occurrence pattern mistakes. */
  public static final MistakeTypeCategory ABSTRACTION_OCCURRENCE_PATTERN_MISTAKES =
      MTCS.get("Abstraction-Occurrence pattern mistakes");


  // Mistake types

  /** The missing class mistake type. */
  public static final MistakeType MISSING_CLASS = MTS.get("Missing class");

  /** The extra class mistake type. */
  public static final MistakeType EXTRA_CLASS = MTS.get("Extra class");

  /** The using n-ary association instead of intermediate class mistake type. */
  public static final MistakeType USING_N_ARY_ASSOC_INSTEAD_OF_INTERMEDIATE_CLASS =
      MTS.get("Using n-ary assoc instead of intermediate class");

  /** The missing attribute mistake type. */
  public static final MistakeType MISSING_ATTRIBUTE = MTS.get("Missing attribute");

  /** The wrong attribute type mistake type. */
  public static final MistakeType WRONG_ATTRIBUTE_TYPE = MTS.get("Wrong attribute type");

  /** The missing attribute type mistake type. */
  public static final MistakeType MISSING_ATTRIBUTE_TYPE = MTS.get("Missing attribute type");

  /** The attribute should be static mistake type. */
  public static final MistakeType ATTRIBUTE_SHOULD_BE_STATIC = MTS.get("Attribute should be static");

  /** The attribute should not be static mistake type. */
  public static final MistakeType ATTRIBUTE_SHOULD_NOT_BE_STATIC = MTS.get("Attribute should not be static");

  /** The plural class name mistake type. */
  public static final MistakeType PLURAL_CLASS_NAME = MTS.get("Plural class name");

  /** The lowercase class name mistake type. */
  public static final MistakeType LOWERCASE_CLASS_NAME = MTS.get("Lowercase class name");

  /** The software engineering term mistake type. */
  public static final MistakeType SOFTWARE_ENGINEERING_TERM = MTS.get("Software engineering term");

  /** The bad class name spelling mistake type. */
  public static final MistakeType BAD_CLASS_NAME_SPELLING = MTS.get("Bad class name spelling");

  /** The wrong class name mistake type. */
  public static final MistakeType WRONG_CLASS_NAME = MTS.get("Wrong class name");

  /** The regular class should be enumeration mistake type. */
  public static final MistakeType CLASS_SHOULD_BE_ENUM = MTS.get("Class should be enum");

  /** The enumeration should be regular class mistake type. */
  public static final MistakeType ENUM_SHOULD_BE_CLASS = MTS.get("Enum should be class");

  /** The missing enumeration mistake type. */
  public static final MistakeType MISSING_ENUM = MTS.get("Missing enum");

  /** The extra enumeration mistake type. */
  public static final MistakeType EXTRA_ENUM = MTS.get("Extra enum");

  /** The bad enumeration name spelling mistake type. */
  public static final MistakeType BAD_ENUM_NAME_SPELLING = MTS.get("Bad enum name spelling");

  /** The missing enumeration item mistake type. */
  public static final MistakeType MISSING_ENUM_ITEM = MTS.get("Missing enum item");

  /** The extra enumeration item mistake type. */
  public static final MistakeType EXTRA_ENUM_ITEM = MTS.get("Extra enum item");

  /** The bad enumeration item spelling mistake type. */
  public static final MistakeType BAD_ENUM_ITEM_SPELLING = MTS.get("Bad enum item spelling");

  /** The plural attribute mistake type. */
  public static final MistakeType PLURAL_ATTRIBUTE = MTS.get("Plural attribute");

  /** The list attribute mistake type. */
  public static final MistakeType LIST_ATTRIBUTE = MTS.get("List attribute");

  /** The extra attribute mistake type. */
  public static final MistakeType EXTRA_ATTRIBUTE = MTS.get("Extra attribute");

  /** The bad attribute name spelling mistake type. */
  public static final MistakeType BAD_ATTRIBUTE_NAME_SPELLING = MTS.get("Bad attribute name spelling");

  /** The uppercase attribute name mistake type. */
  public static final MistakeType UPPERCASE_ATTRIBUTE_NAME = MTS.get("Uppercase attribute name");

  /** The attribute misplaced mistake type. */
  public static final MistakeType ATTRIBUTE_MISPLACED = MTS.get("Attribute misplaced");

  /** The attribute duplicated mistake type. */
  public static final MistakeType ATTRIBUTE_DUPLICATED = MTS.get("Attribute duplicated");

  /** The attribute misplaced in generalization hierarchy mistake type. */
  public static final MistakeType ATTRIBUTE_MISPLACED_IN_GENERALIZATION_HIERARCHY =
      MTS.get("Attribute misplaced in generalization hierarchy");

  /** The missing association mistake type. */
  public static final MistakeType MISSING_ASSOCIATION = MTS.get("Missing association");

  /** The missing aggregation mistake type. */
  public static final MistakeType MISSING_AGGREGATION = MTS.get("Missing aggregation");

  /** The missing n-ary association mistake type. */
  public static final MistakeType MISSING_N_ARY_ASSOCIATION = MTS.get("Missing n-ary association");

  /** The using attribute instead of association mistake type. */
  public static final MistakeType USING_ATTRIBUTE_INSTEAD_OF_ASSOC = MTS.get("Using attribute instead of assoc");

  /** The representing an action with an association mistake type. */
  public static final MistakeType REPRESENTING_ACTION_WITH_ASSOC = MTS.get("Representing action with assoc");

  /** The extra association mistake type. */
  public static final MistakeType EXTRA_ASSOCIATION = MTS.get("Extra association");

  /** The extra aggregation mistake type. */
  public static final MistakeType EXTRA_AGGREGATION = MTS.get("Extra aggregation");

  /** The extra n-ary association mistake type. */
  public static final MistakeType EXTRA_N_ARY_ASSOCIATION = MTS.get("Extra n-ary association");

  /** The using aggregation instead of association mistake type. */
  public static final MistakeType USING_AGGREGATION_INSTEAD_OF_ASSOC = MTS.get("Using aggregation instead of assoc");

  /** The using composition instead of association mistake type. */
  public static final MistakeType USING_COMPOSITION_INSTEAD_OF_ASSOC = MTS.get("Using composition instead of assoc");

  /** The using directed relationship instead of undirected relationship mistake type. */
  public static final MistakeType USING_DIRECTED_RELATIONSHIP_INSTEAD_OF_UNDIRECTED =
      MTS.get("Using directed relationship instead of undirected");

  /** The using undirected relationship instead of directed relationship mistake type. */
  public static final MistakeType USING_UNDIRECTED_RELATIONSHIP_INSTEAD_OF_DIRECTED =
      MTS.get("Using undirected relationship instead of directed");

  /** The using composition instead of aggregation mistake type. */
  public static final MistakeType USING_COMPOSITION_INSTEAD_OF_AGGREGATION =
      MTS.get("Using composition instead of aggregation");

  /** The using binary association instead of n-ary association mistake type. */
  public static final MistakeType USING_BINARY_ASSOC_INSTEAD_OF_N_ARY_ASSOC =
      MTS.get("Using binary assoc instead of n-ary assoc");

  /** The using n-ary association instead of binary association mistake type. */
  public static final MistakeType USING_N_ARY_ASSOC_INSTEAD_OF_BINARY_ASSOC =
      MTS.get("Using n-ary assoc instead of binary assoc");

  /** The using intermediate class instead of n-ary association mistake type. */
  public static final MistakeType USING_INTERMEDIATE_CLASS_INSTEAD_OF_N_ARY_ASSOC =
      MTS.get("Using intermediate class instead of n-ary assoc");

  /** The missing association name mistake type. */
  public static final MistakeType MISSING_ASSOCIATION_NAME = MTS.get("Missing association name");

  /** The bad association name spelling mistake type. */
  public static final MistakeType BAD_ASSOCIATION_NAME_SPELLING = MTS.get("Bad association name spelling");

  /** The infinite recursive dependency mistake type. */
  public static final MistakeType INFINITE_RECURSIVE_DEPENDENCY = MTS.get("Infinite recursive dependency");

  /** The wrong multiplicity mistake type. */
  public static final MistakeType WRONG_MULTIPLICITY = MTS.get("Wrong multiplicity");

  /** The missing multiplicity mistake type. */
  public static final MistakeType MISSING_MULTIPLICITY = MTS.get("Missing multiplicity");

  /** The missing role names mistake type. */
  public static final MistakeType MISSING_ROLE_NAMES = MTS.get("Missing role names");

  /** The role should be static mistake type. */
  public static final MistakeType ROLE_SHOULD_BE_STATIC = MTS.get("Role should be static");

  /** The role should not be static mistake type. */
  public static final MistakeType ROLE_SHOULD_NOT_BE_STATIC = MTS.get("Role should not be static");

  /** The bad role name spelling mistake type. */
  public static final MistakeType BAD_ROLE_NAME_SPELLING = MTS.get("Bad role name spelling");

  /** The wrong role name but correct association mistake type. */
  public static final MistakeType WRONG_ROLE_NAME = MTS.get("Wrong role name");

  /** The missing association class mistake type. */
  public static final MistakeType MISSING_ASSOC_CLASS = MTS.get("Missing assoc class");

  /** The extra association class mistake type. */
  public static final MistakeType EXTRA_ASSOC_CLASS = MTS.get("Extra assoc class");

  /** The bad association class name spelling mistake type. */
  public static final MistakeType BAD_ASSOC_CLASS_NAME_SPELLING = MTS.get("Bad assoc class name spelling");

  /** The association class should be regular class mistake type. */
  public static final MistakeType ASSOC_CLASS_SHOULD_BE_CLASS = MTS.get("Assoc class should be class");

  /** The regular class should be association class mistake type. */
  public static final MistakeType CLASS_SHOULD_BE_ASSOC_CLASS = MTS.get("Class should be assoc class");

  /** The missing composition mistake type. */
  public static final MistakeType MISSING_COMPOSITION = MTS.get("Missing composition");

  /** The extra composition mistake type. */
  public static final MistakeType EXTRA_COMPOSITION = MTS.get("Extra composition");

  /** The using association instead of aggregation mistake type. */
  public static final MistakeType USING_ASSOC_INSTEAD_OF_AGGREGATION = MTS.get("Using assoc instead of aggregation");

  /** The using association instead of composition mistake type. */
  public static final MistakeType USING_ASSOC_INSTEAD_OF_COMPOSITION = MTS.get("Using assoc instead of composition");

  /** The using aggregation instead of composition mistake type. */
  public static final MistakeType USING_AGGREGATION_INSTEAD_OF_COMPOSITION =
      MTS.get("Using aggregation instead of composition");

  /** The composed part contained in more than one parent mistake type. */
  public static final MistakeType COMPOSED_PART_CONTAINED_IN_MORE_THAN_ONE_PARENT =
      MTS.get("Composed part contained in more than one parent");

  /** The incomplete containment tree mistake type. */
  public static final MistakeType INCOMPLETE_CONTAINMENT_TREE = MTS.get("Incomplete containment tree");

  /** The missing generalization mistake type. */
  public static final MistakeType MISSING_GENERALIZATION = MTS.get("Missing generalization");

  /** The extra generalization mistake type. */
  public static final MistakeType EXTRA_GENERALIZATION = MTS.get("Extra generalization");

  /** The generalization does not follow isA rule mistake type. */
  public static final MistakeType GENERALIZATION_INAPPLICABLE = MTS.get("Generalization inapplicable");

  /** The subclass not distinct across lifetime mistake type. */
  public static final MistakeType SUBCLASS_NOT_DISTINCT_ACROSS_LIFETIME =
      MTS.get("Subclass not distinct across lifetime");

  /** The inherited feature does not make sense for subclass mistake type. */
  public static final MistakeType INHERITED_FEATURE_DOES_NOT_MAKE_SENSE_FOR_SUBCLASS =
      MTS.get("Inherited feature does not make sense for subclass");

  /** The subclass is an instance of superclass mistake type. */
  public static final MistakeType SUBCLASS_IS_AN_INSTANCE_OF_SUPERCLASS =
      MTS.get("Subclass is an instance of superclass");

  /** The non-differentiated subclass mistake type. */
  public static final MistakeType NON_DIFFERENTIATED_SUBCLASS = MTS.get("Non-differentiated subclass");

  /** The wrong generalization direction mistake type. */
  public static final MistakeType WRONG_GENERALIZATION_DIRECTION = MTS.get("Wrong generalization direction");

  /** The wrong superclass mistake type. */
  public static final MistakeType WRONG_SUPERCLASS = MTS.get("Wrong superclass");

  /** The subclass should be full player-role pattern mistake type. */
  public static final MistakeType SUBCLASS_SHOULD_BE_FULL_PR_PATTERN = MTS.get("Subclass should be full PR pattern");

  /** The subclass should be association player-role pattern mistake type. */
  public static final MistakeType SUBCLASS_SHOULD_BE_ASSOC_PR_PATTERN = MTS.get("Subclass should be assoc PR pattern");

  /** The subclass should be enumeration player-role pattern mistake type. */
  public static final MistakeType SUBCLASS_SHOULD_BE_ENUM_PR_PATTERN = MTS.get("Subclass should be enum PR pattern");

  /** The association should be full player-role pattern mistake type. */
  public static final MistakeType ASSOC_SHOULD_BE_FULL_PR_PATTERN = MTS.get("Assoc should be full PR pattern");

  /** The association should be subclass player-role pattern mistake type. */
  public static final MistakeType ASSOC_SHOULD_BE_SUBCLASS_PR_PATTERN = MTS.get("Assoc should be subclass PR pattern");

  /** The association should be enumeration player-role pattern mistake type. */
  public static final MistakeType ASSOC_SHOULD_BE_ENUM_PR_PATTERN = MTS.get("Assoc should be enum PR pattern");

  /** The enumeration should be full player-role pattern mistake type. */
  public static final MistakeType ENUM_SHOULD_BE_FULL_PR_PATTERN = MTS.get("Enum should be full PR pattern");

  /** The enumeration should be subclass player-role pattern mistake type. */
  public static final MistakeType ENUM_SHOULD_BE_SUBCLASS_PR_PATTERN = MTS.get("Enum should be subclass PR pattern");

  /** The enumeration should be association player-role pattern mistake type. */
  public static final MistakeType ENUM_SHOULD_BE_ASSOC_PR_PATTERN = MTS.get("Enum should be assoc PR pattern");

  /** The full player-role pattern should be subclass mistake type. */
  public static final MistakeType FULL_PR_PATTERN_SHOULD_BE_SUBCLASS = MTS.get("Full PR pattern should be subclass");

  /** The full player-role pattern should be association mistake type. */
  public static final MistakeType FULL_PR_PATTERN_SHOULD_BE_ASSOC = MTS.get("Full PR pattern should be assoc");

  /** The full player-role pattern should be enumeration mistake type. */
  public static final MistakeType FULL_PR_PATTERN_SHOULD_BE_ENUM = MTS.get("Full PR pattern should be enum");

  /** The missing player-role pattern mistake type. */
  public static final MistakeType MISSING_PR_PATTERN = MTS.get("Missing PR pattern");

  /** The incomplete player-role pattern mistake type. */
  public static final MistakeType INCOMPLETE_PR_PATTERN = MTS.get("Incomplete PR pattern");

  /** The missing abstraction-occurrence pattern mistake type. */
  public static final MistakeType MISSING_AO_PATTERN = MTS.get("Missing AO pattern");

  /** The incomplete abstraction-occurrence pattern mistake type. */
  public static final MistakeType INCOMPLETE_AO_PATTERN = MTS.get("Incomplete AO pattern");

  /** The generalization should be association in abstraction-occurrence pattern mistake type. */
  public static final MistakeType GENERALIZATION_SHOULD_BE_ASSOC_AO_PATTERN =
      MTS.get("Generalization should be assoc AO pattern");

}
